### PR TITLE
Switching pillars to OrderedDicts for salt-master and salt-ssh

### PR DESCRIPTION
**Pull request by edgan**
### What does this PR do?
It makes pillars OrderedDicts for salt-master and salt-ssh.

### What issues does this PR fix or reference?
#42440

### Previous Behavior
The order of pillars a, b, c become  b, c, a.

### New Behavior
Pillars a, b, c stay in a, b, c order.

### CODE REVIEW

**Review by terminalmage — CHANGES_REQUESTED** [review]

File: salt/pillar/__init__.py [hunk 1]
```diff
@@ -145,10 +145,10 @@ def __init__(self, opts, grains, minion_id, saltenv, ext=None, functions=None,
         self.channel = salt.transport.client.AsyncReqChannel.factory(opts)
         if pillarenv is not None:
             self.opts['pillarenv'] = pillarenv
-        self.pillar_override = pillar_override or {}
-        if not isinstance(self.pillar_override, dict):
-            self.pillar_override = {}
-            log.error('Pillar data must be a dictionary')
+        self.pillar_override = pillar_override or OrderedDict()
+        if not isinstance(self.pillar_override, dict) and not isinstance(self.pillar_override, OrderedDict):
```
Comment (terminalmage) [inline salt/pillar/__init__.py]: OrderedDict inherits from `dict`, the second isinstance check is redundant:

```python
>>> import salt.utils.odict
>>> isinstance(salt.utils.odict.OrderedDict(), dict)
True
```

I'm also confused by your recent commit. I thought you wanted pillar to be an OrderedDict, why did you suddenly change this?

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by rallytime** [comment]
@edgan There's a small lint error here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/19731/violations/file/salt/pillar/__init__.py/

**Comment by edgan** [comment]
The lint error data is MIA. Can you point me to the lint problem?

**Comment by gtmanfred** [comment]
re-run lint

We have to rotate logs so we don't run out of space on the salt master.  Also, looks like you will need a rebase to fix a conflict in salt/payload.py

**Comment by rallytime** [comment]
You can also run the lint check locally. [Here's some docs.](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#linting) :D

**Comment by terminalmage** [comment]
@edgan Can you respond to my question here? https://github.com/saltstack/salt/pull/46257#discussion_r171455465

**Comment by edgan** [comment]
I thought I had. I had been talking to someone else who had suggested the change. I have since removed it. I had the same comment to him that OrderedDicts match as dicts.

**Review by terminalmage — APPROVED** [review]

**Comment by rallytime** [comment]
re-run py

**Comment by rallytime** [comment]
I am seeing this test fail here, but not on the branch. I think it is related to this change:

- integration.shell.test_cp.CopyTest.test_cp_testfile

**Comment by rallytime** [comment]
@edgan Can you take a look at these test failures?

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/8380/

**Comment by rallytime** [comment]
The tests that are failing due to this change are the following:

- unit.test_payload.PayloadTestCase.test_list_nested_odicts
- integration.shell.test_cp.CopyTest.test_cp_testfile

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23938/

@edgan Can you take a look? We won't be able to merge this until those are addressed.

**Comment by rallytime** [comment]
Since we haven't heard back here for a while, I am going to close this. Once those tests are addressed, we can certainly open this again. Or feel free to open a new PR, rebased on the latest develop. Thank you!

**Comment by ixs** [comment]
@edgan do you think you could revisit this change? Having it merged would actually be very nice.

**Comment by wilminator** [comment]
I will chime in because I proposed a similar fix in payload.py.  The Payload test case checks specifically if an OrderedDict comes out of the payload unserializing methods.  Because this fix causes that to happen, PayloadTestCase errors.

I asked specifically about the purpose for that test, but have not heard any response about it.  The unit test only indicates what it is checking for.  I don't think there would be any problem with a dict retaining order; it is highly unlikely that any code in this project relies on the randomization of the key order of the dict object prior to Python 3.6.

With that said, can anyone see any reason to keep that test, or would it be acceptable for me to commit the ommission of that test?

I will further look into the CopyTest fail.

**Comment by wilminator** [comment]
Okay, it makes sense.  I was reviewing test_payload.py and it finally hit me.  This unit test intentionally anticipates data being unserialized back into primitive object types (tuples, lists, and dicts) from msgpack.  The payload.py modification changes that to tuples, lists, and OrderedDicts, unless the msgpack version id too old (<0.4.0).

Would it be acceptable because of the intention of this change to alter the unit test to accommodate for an OrderedDict output instead of a dict in the payload?  Any suggestions how to edit the test code so we can test for both when msgpack is lesser or newer than 0.4.0? I would happily do that work if pointed in the right direction.  I need just a bit more direction than the "write a test" tutorials; I think i get the concept but am not sure how to fake the version value of msgpack.

**Comment by wilminator** [comment]
Based on #46868 and saltstack/salt-pack#537 no one should be using msgpack < 0.4.0.  In addition, object_pairs_hooks exists back to at least msgpack 0.2.4.  I can update the payload and test code accordingly.

**Comment by ixs** [comment]
@rallytime @gtmanfred Any ideas about the proposed fix from @wilminator?

**Comment by wilminator** [comment]
@ixs to be fair, I have described a way to move forward but have not proposed any fix- there is no pull request by anybody.  Give me and others a little more time to submit a pull request first.

**Pull request closed without merging**